src/components/SnackChat/SnackChat.js

```diff
@@ -78,112 +59,140 @@ class SnackChat extends Component {
+          FALLING_SNACK_SIZE * FEED_SIZE;
         item.rotation = (item.rotation + 0.03) % (Math.PI * 2);
       });
-
       requestAnimationFrame(drawFallingSnacks);
     };
-
     requestAnimationFrame(drawFallingSnacks);
-    this.ctx.clearRect(
-      0,
-      0,
-      this.canvas.current.width,
-      this.canvas.current.height
+  };
+
+  clipEllipse(ctx, centerX, centerY, width, height) {
+    ctx.beginPath();
+    ctx.moveTo(centerX, centerY - height / 2);
+    ctx.bezierCurveTo(
+      centerX + width / 2,
+      centerY - height / 2,
+      centerX + width / 2,
+      centerY + height / 2,
+      centerX,
+      centerY + height / 2
+    );
+    ctx.bezierCurveTo(
+      centerX - width / 2,
+      centerY + height / 2,
+      centerX - width / 2,
+      centerY - height / 2,
+      centerX,
+      centerY - height / 2
+    );
+    ctx.clip();
+  }
+
+  drawBackground = video => {
+    this.ctx.scale(-CANVAS_SCALE, CANVAS_SCALE);
+    this.ctx.drawImage(
+      video,
+      Math.abs(video.videoWidth - this.canvas.current.width) / 2 -
+        video.videoWidth,
+      0
     );
+    this.ctx.restore();
+  };
+
+  generateSnackchat = async () => {
+    const video = this.webcam.current.webcam.current.video;
+
+    this.ctx.save();
+    this.drawBackground(video);
+
+    const filter = new Image();
+    filter.src = this.filter;
+
+    const shoulders = this.state.averageBodyPosition.shoulders;
+    this.ctx.save();
+    this.ctx.drawImage(
+      filter,
+      shoulders.rightX -
+        shoulders.span * 1.5 +
+        shoulders.span * shoulders.angle,
+      shoulders.rightY - shoulders.span * 1.5,
+      shoulders.span * 4,
+      shoulders.span * 4
+    );
+    this.ctx.restore();
+
+    const ears = this.state.averageBodyPosition.ears;
+    this.ctx.save();
+    this.ctx.translate(
+      ears.rightX + ears.width / 2,
+      ears.rightY + ears.height * ears.angle
+    );
+    this.ctx.rotate(ears.angle);
+    this.clipEllipse(this.ctx, 0, 0, ears.span * 1.5, ears.span * 1.5);
+    this.ctx.resetTransform();
+
+    this.drawBackground(video);
   };
 
   positionBuffer = new Array(POSITION_BUFFER_SIZE);
-  averageBodyPosition;
   i = -1;
+  redirected = false;
   update = async () => {
-    if (!this.webcam.current.webcam.current || !this.net) {
+    if (this.redirected || !this.webcam.current.webcam.current || !this.net) {
       requestAnimationFrame(this.update);
       return;
     }
 
     if (this.state.counter <= PHOTO_ANIMATION_TIME && !this.state.captured) {
-      setTimeout(() => {
-        this.setState({captured: true});
-        this.props.setSnackChat(this.canvas.current.toDataURL());
-        document.getElementById('fade-overlay').className = 'fade-in';
-        setTimeout(() => {
-          clearInterval(this.timer);
-          this.props.history.replace('/slackname');
-        }, 1000);
-      }, (PHOTO_ANIMATION_TIME + 1) * 1000);
+      // async so that the filter doesn't stop moving
```

Comment:
Since we're using the value of state to potentially update state we'll want to use the updater form of `setState` since `this.state` could be out of date

It'd be good to extract the setTimeout callback to a descriptive method too

```
this.setState((prevState) => {
  if (prevState.counter <= PHOTO_ANIMATION_TIME && !prevState.captured) {
    setTimeout(this.sendSnackchat, timeout)

    return { captured: true }
  }

  return null
})
```